janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/MultiThreadsScanner.java

```diff
@@ -0,0 +1,259 @@
+// limitations under the License.
+
+package org.janusgraph.diskstorage.keycolumnvalue.scan;
+
+import org.janusgraph.diskstorage.BackendException;
+import org.janusgraph.diskstorage.Entry;
+import org.janusgraph.diskstorage.EntryList;
+import org.janusgraph.diskstorage.StaticBuffer;
+import org.janusgraph.diskstorage.TemporaryBackendException;
+import org.janusgraph.diskstorage.configuration.Configuration;
+import org.janusgraph.diskstorage.keycolumnvalue.KCVSUtil;
+import org.janusgraph.diskstorage.keycolumnvalue.KeyColumnValueStore;
+import org.janusgraph.diskstorage.keycolumnvalue.KeyIterator;
+import org.janusgraph.diskstorage.keycolumnvalue.SliceQuery;
+import org.janusgraph.diskstorage.keycolumnvalue.StoreFeatures;
+import org.janusgraph.diskstorage.keycolumnvalue.StoreTransaction;
+import org.janusgraph.diskstorage.util.EntryArrayList;
+import org.janusgraph.diskstorage.util.RecordIterator;
+import org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+
+import static org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor.Row;
+import static org.janusgraph.diskstorage.keycolumnvalue.scan.StandardScannerExecutor.TIME_PER_TRY;
+
+/**
+ * @author Sergii Karpenko (sergiy.karpenko@gmail.com)
+ */
+
+class MultiThreadsScanner extends Scanner {
+
+    private static final int MAX_KEY_LENGTH = 128; //in bytes
+
+    private static final Logger log = LoggerFactory.getLogger(MultiThreadsScanner.class);
+
+    private final KeyColumnValueStore store;
+    private final StoreFeatures storeFeatures;
+    private final StoreTransaction storeTx;
+    private final List<SliceQuery> queries;
+    private final Predicate<StaticBuffer> keyFilter;
+    private final Configuration graphConfiguration;
+    private final List<DataPuller> pullThreads;
+    private final List<BlockingQueue<SliceResult>> dataQueues;
+    private boolean interrupted = false;
+
+    MultiThreadsScanner(
+        KeyColumnValueStore store,
+        StoreFeatures storeFeatures,
+        StoreTransaction storeTx,
+        List<SliceQuery> queries,
+        Predicate<StaticBuffer> keyFilter,
+        BlockingQueue<Row> rowQueue,
+        Configuration graphConfiguration) throws BackendException {
+
+        super(store, rowQueue);
+        this.store = store;
+        this.storeFeatures = storeFeatures;
+        this.storeTx = storeTx;
+        this.queries = queries;
+        this.keyFilter = keyFilter;
+        this.graphConfiguration = graphConfiguration;
+
+        this.dataQueues = new ArrayList<>(queries.size());
+        this.pullThreads = new ArrayList<>(queries.size());
```

Comment:
(nitpick) Previously it was an array and not a collection. Is there a reason why you changed it from non resizable array to resizable array list?

Reply:
reverted from List to array